tools/update-index/Program.cs

```diff
@@ -0,0 +1,263 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+using Mono.Options;
+
+internal static class Program
+{
+    private static int Main(string[] args)
+    {
+        var exeName = Path.GetFileNameWithoutExtension(Environment.GetCommandLineArgs()[0]);
+        var help = false;
+        var outputPath = "";
+        var directoryPath = "";
+
+        var options = new OptionSet
+        {
+            $"usage: {exeName} <directory> [OPTIONS]+",
+            { "o|out=", "The output {path} where the index should be written to. Default: <directory>/INDEX.md", v => outputPath = v },
+            { "h|?|help", null, v => help = true, true },
+            new ResponseFileSource()
+        };
+
+        try
+        {
+            var parameters = options.Parse(args).ToArray();
+
+            if (help)
+            {
+                options.WriteOptionDescriptions(Console.Error);
+                return 0;
+            }
+
+            if (parameters.Length >= 1)
+            {
+                directoryPath = parameters[0];
+            }
+            else
+            {
+                Console.Error.WriteLine("error: must specify a directory");
+                return 1;
+            }
+
+            var unprocessed = parameters.Skip(1);
+
+            if (unprocessed.Any())
+            {
+                foreach (var option in unprocessed)
+                {
+                    Console.Error.WriteLine($"error: unrecognized argument {option}");
+                }
+
+                return 1;
+            }
+        }
+        catch (Exception ex)
+        {
+            Console.Error.WriteLine(ex);
+            return 1;
+        }
+
+        if (!Directory.Exists(directoryPath))
+        {
+            Console.Error.WriteLine($"error: directory '{directoryPath}' does not exist");
+            return 1;
+        }
+
+        if (string.IsNullOrEmpty(outputPath))
+        {
+            outputPath = Path.Join(directoryPath, "INDEX.md");
+        }
+
+        try
+        {
+            UpdateIndex(directoryPath, outputPath);
+            return 0;
+        }
+        catch (Exception ex)
+        {
+            Console.Error.WriteLine(ex);
+            return 1;                
+        }
+    }
+
+    private static void UpdateIndex(string directoryPath, string outputPath)
+    {
+        var designs = Directory.GetFiles(directoryPath, "*.md", SearchOption.AllDirectories)
+                               .Select(p => Document.Parse(p))
+                               .Where(d => d != null)
+                               .Select(d => d!)
+                               .ToArray();
+
+        using var outputWriter = new StreamWriter(outputPath);
+
+        outputWriter.WriteLine("<!--");
```

Comment:
Not sure the current state of T4 tooling, but this would be more readable as a T4 template.

Reply:
I just tried it. Couple of things:

1. It doesn't work well without VS because the generated T4 file is only updated when you're within VS. If you're doing VS Code and CLI, then I don't even know how you can re-generate the file.
2. Suprisingly T4 seems to work just fine .NET 5
3. Razor is much nicer than T4 because you don't have to declare the transitions between code and markup. In T4 you need to use markers and you need to be very careful with line breaks in Markdown. For prosperity I've included the T4 version below. TBH, I don't find it more readable.

So yeah, I'm inclined to not use T4 here.

```C#
<#@ template language="C#" #>
<#@ assembly name="System.Core" #>
<#@ import namespace="System.Linq" #>
<#@ import namespace="System.Text" #>
<#@ import namespace="System.Collections.Generic" #>
<!--

This file is auto-generated. Direct changes to it maybe lost.

Use update-index to regenerate it:

    ./update-index

-->

# Design Index

## Meta

<#
foreach (var document in GetDocuments(DocumentKind.Meta)) { #>
* <#= GetMarkdownLink(document) #>
<#
}
#>

## Accepted

|Year|Title|Owners|
|----|-----|------|
<#
foreach (var document in GetDocuments(DocumentKind.AcceptedDesign))
{
    var year = document.Year.ToString();
    var link = GetMarkdownLink(document);
    var owners = string.Join(", ", document.Owners);
#>| <#= year #> | <#= link #> | <#= owners #> |
<#
}
#>

## Proposed

|Year|Title|Owners|
|----|-----|------|
<#
foreach (var document in GetDocuments(DocumentKind.ProposedDesign))
{
    var year = document.Year.ToString();
    var link = GetMarkdownLink(document);
    var owners = string.Join(", ", document.Owners);
#>| <#= year #> | <#= link #> | <#= owners #> |
<#
}
#>


```